pkg/providers/vsphere/upgrade/virtualmachine/vm_schema_upgrade.go

```diff
@@ -397,39 +406,358 @@ func reconcileSCSIController(
+	}
+
+	var (
+		cdromDevices []*vimtypes.VirtualCdrom
+		diskDevices  []*vimtypes.VirtualDisk
+	)
+
 	for i := range moVM.Config.Hardware.Device {
 		switch d := moVM.Config.Hardware.Device[i].(type) {
 
 		case *vimtypes.VirtualDisk:
-			reconcileVirtualDisk(ctx, vm, d)
+			diskDevices = append(diskDevices, d)
 
 		case *vimtypes.VirtualCdrom:
-			reconcileVirtualCDROM(ctx, vm, d)
+			cdromDevices = append(cdromDevices, d)
+		}
+	}
+
+	reconcileVirtualCDROMs(ctx, k8sClient, vm, ctrlKeyToCtrlStatus, cdromDevices)
+	reconcileVirtualDisks(ctx, k8sClient, vm, ctrlKeyToCtrlStatus, diskDevices)
+}
+
+// reconcileVirtualDisks reconciles the VM's disk devices during
+// schema upgrade. It maps disk devices from the vSphere VM
+// configuration to the VirtualMachine spec volumes, updating the
+// unit number, controller type, and controller bus number for each
+// PVC volume based on the disk UUID and controller information.
+func reconcileVirtualDisks(
+	ctx context.Context,
+	k8sClient ctrlclient.Client,
+	vm *vmopv1.VirtualMachine,
+	ctrlKeyToCtrlStatus map[int32]vmopv1.VirtualControllerStatus,
+	diskDevices []*vimtypes.VirtualDisk) {
+
+	logger := pkglog.FromContextOrDefault(ctx)
+	logger.V(4).Info("Reconciling schema upgrade for VM disks")
+
+	if vm.Status.BiosUUID == "" {
+		logger.V(4).Info(
+			"Skipping PVC volume backfilling, BiosUUID not available")
+		return
+	}
+
+	diskUUIDToDiskInfo := buildDiskUUIDToDiskInfoMap(diskDevices)
+	volumeNameToDiskUUID, err := buildVolumeNameToDiskUUIDMap(
+		ctx, k8sClient, vm)
+	if err != nil {
+		return
+	}
+
+	for i := range vm.Spec.Volumes {
+		vol := &vm.Spec.Volumes[i]
+
+		if vol.PersistentVolumeClaim == nil {
+			continue
 		}
+
+		reconcilePVCVolumePlacement(
+			ctx,
+			vol,
+			volumeNameToDiskUUID,
+			diskUUIDToDiskInfo,
+			ctrlKeyToCtrlStatus)
 	}
 }
 
-func reconcileVirtualDisk(
+// buildDiskUUIDToDiskInfoMap builds a map of disk UUID to disk
+// device info by extracting from vSphere VM hardware.
+func buildDiskUUIDToDiskInfoMap(
+	diskDevices []*vimtypes.VirtualDisk) map[string]pkgutil.VirtualDiskInfo {
+
+	diskUUIDToDiskInfo := make(map[string]pkgutil.VirtualDiskInfo)
+	for _, disk := range diskDevices {
+		vdi := pkgutil.GetVirtualDiskInfo(disk)
+		if vdi.UUID != "" {
+			diskUUIDToDiskInfo[vdi.UUID] = vdi
+		}
+	}
+	return diskUUIDToDiskInfo
+}
+
+// buildVolumeNameToDiskUUIDMap builds a map of volume name to
+// disk UUID by querying CnsNodeVmAttachment objects. This is the
+// source of truth for PVC volume attachments.
+func buildVolumeNameToDiskUUIDMap(
 	ctx context.Context,
-	_ *vmopv1.VirtualMachine,
-	_ *vimtypes.VirtualDisk) {
+	k8sClient ctrlclient.Client,
+	vm *vmopv1.VirtualMachine) (map[string]string, error) {
 
 	logger := pkglog.FromContextOrDefault(ctx)
-	logger.V(4).Info("Reconciling schema upgrade for VM disk")
+	volumeNameToDiskUUID := make(map[string]string)
+	attachments, err := pkgutil.GetAttachmentsForVM(ctx, k8sClient, vm)
+	if err != nil {
+		logger.V(4).Info(
```

Comment:
Is this error not logged at all a caller? Can we wrap the error instead?

BTW - it is weird to use logger.Info(.., "err", err) instead of logger.Error()